File: 3170-lexicographically-minimum-string-after-removing-stars/3170-lexicographically-minimum-string-after-removing-stars.cpp
class Solution {
public:
    string clearStars(string s) {
        int n = s.size();
        priority_queue<char, vector<char>, greater<char>> pq;  // Min-heap
        unordered_map<char, vector<int>> m;  // Stores indices of each character
        vector<bool> v(n, true);  // Marks whether each char in s should be kept

        for (int i = 0; i < n; i++) {
            if (s[i] == '*') {
                char temp = pq.top();  // Get smallest char so far
                pq.pop();
                int last = m[temp].back();  // Get last occurrence index
                m[temp].pop_back();
                v[i] = false;       // Remove '*'
                v[last] = false;    // Remove corresponding smallest char
            } else {
                pq.push(s[i]);
                m[s[i]].push_back(i);
            }
        }

        string result = "";
        for (int i = 0; i < n; i++) {
            if (v[i]) result += s[i];
        }
        return result;
    }
};

File: 3170-lexicographically-minimum-string-after-removing-stars/3170-lexicographically-minimum-string-after-removing-stars.py
import heapq
from collections import defaultdict, deque

class Solution(object):
    def clearStars(self, s):
        n = len(s)
        pq = []  # Min-heap to store characters
        m = defaultdict(deque)  # Map character -> deque of indices
        keep = [True] * n  # Marks whether to keep character at index

        for i in range(n):
            if s[i] == '*':
                smallest = heapq.heappop(pq)  # Get smallest character
                idx = m[smallest].pop()      # Remove last occurrence index
                keep[i] = False              # Mark '*' for removal
                keep[idx] = False            # Mark smallest char for removal
            else:
                heapq.heappush(pq, s[i])
                m[s[i]].append(i)

        # Build result from characters not marked for removal
        return ''.join(s[i] for i in range(n) if keep[i])